### PR TITLE
Fix for StackOverflow in SsaBuilder::TopologicalSort.

### PR DESCRIPTION
**Pull request by erozenfeld**
SsaBuilder::TopologicalSort allocated an array of bools on the stack.
In one of internal cases the array had 126,758 elements, which caused a
StackOverflow. This change switches SsaBuilder::TopologicalSort from
stack-allocated array of bools to BitVec.

TopologicalSortHelper used to be recursive. It was changed to use an iterative
algorithm in a previous change. I'm inlining TopologicalSortHelper into
TopologicalSort now.

I verified no measurable negative throughput impact from this change
and no asm diffs in SuperPMI.

The changes were mostly implemented by @russellhadley.

@dotnet/jit-contrib PTAL

### CODE REVIEW

**Comment by cmckinsey** [comment]
LGTM. If someone allocated data structure on the stack as function of input flow graph, they might do it again. Worth a regression test (which apparently we must have a hole today?)

**Comment by BruceForstall** [comment]
LGTM

File: src/jit/ssabuilder.cpp [hunk 1]
```diff
@@ -232,17 +151,85 @@ SsaBuilder::SsaBuilder(Compiler* pCompiler, IAllocator* pIAllocator)
  */
 int SsaBuilder::TopologicalSort(BasicBlock** postOrder, int count)
```
Comment (CarolEidt) [inline src/jit/ssabuilder.cpp]: It would be nice to update the function header according to the coding conventions.

**Comment by CarolEidt** [comment]
:shipit:

File: src/jit/ssabuilder.cpp (hunk 1 above)
Comment (erozenfeld) [inline src/jit/ssabuilder.cpp]: Done.

**Comment by russellhadley** [comment]
LGTM - Thanks for shepherding these changes through.

**Comment by erozenfeld** [comment]
@cmckinsey I added a regression test that causes StackOverflow with a desktop crossgen (without the fix). It doesn't cause a StackOverflow with CoreCLR crossgen because the stack size is 1.8x larger for CoreCLR crossgen. When I tried to make the test larger I ran into issues with lvRefCnt and csdUseCount overflowing and causing asserts. I'll open an issue to track that.

**Comment by erozenfeld** [comment]
I also noticed that SsaBuilder::Build had an alloca of an array whose size was equal to the number of basic blocks. I fixed that by stack allocating the array only if the number of basic blocks is not greater than JitConfig.JitMinOptsBbCount() and heap allocating the array otherwise.

**Comment by erozenfeld** [comment]
@cmckinsey @russellhadley PTAL

**Comment by erozenfeld** [comment]
I opened #7277 to track  lvRefCnt and csdUseCount overflows.

**Comment by russellhadley** [comment]
LGTM

**Comment by erozenfeld** [comment]
I added the new test to the list of exclusions on x86 since it's causing OutOfMemory on x86.

**Pull request merged**